nullaway/src/test/java/com/uber/nullaway/NullAwaySerializationTest.java

```diff
@@ -383,7 +383,7 @@ public void suggestNullableParamGenericsTest() {
         .setExpectedOutputs(
             new FixDisplay(
                 "nullable",
-                "newStatement(T,java.util.ArrayList<T>,boolean,boolean)",
+                "newStatement(T,java.util.ArrayList,boolean,boolean)",
```

Comment:
Since generics are erased at the bytecode level, the `T` here would turn into `java.lang.Object` in a JVM-level method descriptor.  But, as long as we also serialize the enclosing class, I think keeping `T` here should be no problem.  @nimakarimipour do you agree?  Assuming yes, we are fine

Reply:
@msridhar Sorry, I am not sure if I understood correctly. Currently we are not serializing type arguments, therefore, for `ArrayList<T>` in a method, we only serialize `java.util.ArrayList`. Our current implementation does not visit the type arguments at all. If we decide to keep `T` in this case. I am not sure how we should serialize `Map<Object, ArrayList<T>>` without full serialization of all type arguments which requires switching to visitor. Our current implementation in this PR will only serialize `java.util.Map`.

Considering Annotator, for a method serialization we require two conditions.

1. A method serialization which is robust to annotation injection / deletion in source code.
2. Preserve enough information in the string representation so that Injector can locate the exact method declaration.

I think `java.util.ArrayList` or `java.util.Map` satisfies both conditions. Would you please elaborate what I should do or what I am missing here. Thank you.

Reply:
I think Manu's question here is just, why the signature isn't simply:

```
newStatement(java.lang.Object,java.util.ArrayList,boolean,boolean)
```

if `T` is type-erased to `java.lang.Object`.

The problem is that if we want to perform type erasure of arguments with a type argument as their type, then we need to be very careful. And we need to be careful here and also in the javaparser code, no?

For example, here the compiler will throw an error, because indeed, at the JVM level, these two `bar` methods are one and the same:

```
    class Foo<T> {
        public void bar(T t) { }
        public void bar(Object t) { }
    }
```

However, simply changing the code to:

```
    class Foo<T extends String> {
        public void bar(T t) { }
        public void bar(Object t) { }
    }
```

will caused to compile. Because now the JVM signature of the first `bar` is `bar(java.lang.String)`, which doesn't clash with `bar(java.lang.Object)` below.

I think if both NullAway and the javaparser code are looking at the source and serializing `"T"`, then that's the most robust case. But a question is, don't we serialize methods from bytecode jars when looking at the downstream dependencies? Is there a case where serialization is looking at a symbol from bytecode, yet we are then trying to resolve the method name to inject an annotation?

p.s. presumably getting the `Signature.descriptor` already returns the type-erased signature?

Reply:
Yes @lazaroclapp that's what I was thinking about.  And I think if we serialize `T` in _all_ cases we are ok.  But you're right that if we need to serialize the same method signature both from source code and also from bytecode then that could be tricky.  @nimakarimipour what do you think?

I presume `Signature.descriptor` returns a type-erased signature but I haven't tested to see if it does the right thing for the cases above (I actually didn't even though that the upper bound of a type variable can affect the generated bytecode!)

Reply:
@msridhar @lazaroclapp Sorry for the confusion and thank you for the clarification, now I understand the issue. I just tested the serialization. It serializes `T` in _all_ cases. I applied the same logic here for method serialization in annotator-scanner and checked for downstream dependencies, for a method with generic type in its argument, it can correctly:
1. Scan the potentially impacted regions on downstream dependencies.
2. Serialize a fix for an element in upstream.

Please see the example used in test:
```java
// In upstream
Foo<T> {
  public T bar(T p1) {
       return null;
  }
}
```
and 
```java
// In downstream
Dep {
  Foo<Object> foo;
  public T baz() {
       return foo.bar(null);
  }
}
```


For method `Foo#bar(T)` it can correctly detect `Dep#baz()` is an impacted region and the fix for passing `null` in `Dep#baz()` is making `Foo#bar(T)` accept nullable. 

An interesting point is that parameter names are not consistent for `Foo` while processing downstream dependencies since received as jar. We weren't depending on parameter names and were using index, but I should just keep in mind never depend on that for future works. 🙂.

So I think we can keep the serialization as is.

Reply:
Shouldn't `Dep#baz()` return `Object` rather than `T`?  But otherwise this makes sense to me.  At runtime, for backward compatibility with pre-generics bytecodes, generic types are erased in method descriptors.  But, all the generic type information still needs to be preserved _somewhere_ in the class files, to enable proper type checking of downstream deps in `javac`.  Fortunately, it seems that `javac` returns the proper generic type info for upstream for the APIs we are using.

Reply:
@msridhar Yes, exactly. I simplified the actual tested source code and missed this. It was `Object` in `Dep#baz()`.

Reply:
So we get un-erased generics even when looking at jar files? Cool!